### DISCUSSION TITLE
"Opt out" for metrics collection

### DISCUSSION THREAD

**Discussion by jwalton** [cds-snc/covid-alert-documentation#41] [Ideas] [answered]
# Summary | Résumé

As a user, I want to be able to opt out of metrics collection, because I don't want privacy-conscious users to uninstall this app.

# Design detail | Précisions au sujet de la conception

There should be a clear "Collect anonymous usage data" toggle in the app settings.  The app should ask the user for consent to collect anonymous data about them before enabling this feature.

The new metrics collection feature was just rolled out, but absent an "opt out" setting, the only way for privacy-conscious users to opt out is to uninstall the app entirely.  While it's understandable that CDS-SNC wants anonymous usage statistics, it's also understandable that some users may not want to supply them.  It would be better for CDS-SNC to only get anonymous statistics from some of their users, as opposed to gettings stats from all users but have less users overall.

**Comment by timarney** [comment]
cc: @emilykuret @sidewalkballet

**Comment by obrien-j** [comment]
Re-homing this to the central covid-alert-documentation repo, inline with the new rfc 0001-working-in-the-open

**Comment by obrien-j** [comment, marked as answer]
Thanks for the submission @jwalton,

As you mentioned, this change was rolled out to be able to gather aggregate numbers so that we can enhance the Government of Canada’s ability to evaluate the effectiveness of COVID Alert while still preserving the app’s privacy-protecting approach. 

For example, this information will help us assess how many cases of COVID-19 might have been avoided due to the app. We will also be able to use this information to encourage uptake by demonstrating to the public, with tangible data, the benefits of downloading the app. 

On the technical side, having more information on how the app is functioning will help us better identify and respond to bugs or other issues. This will enable us to respond more quickly to issues facing the app, better prioritize bug fixes according to their scope and ensure that the app is acting as intended. 

Your point is valid, and we will continue to monitor feedback from users and work with provincial and territorial partners to learn and understand if people are choosing to delete the app. The voluntary nature of the app means there is always a possibility that app users will delete the app.

I’ll leave this idea open in case others have thoughts on the matter.

**Reply by jwalton** [reply]
> ...to be able to gather aggregate numbers

The data that's actually sent from my phone is a notification that I was shown, for example, a "You have been exposed to covid" alert, and my IP address.  With my IP address, CDS-SNC receives all the information they need to track whom has and has not been shows such a notification.  Users only have your word that you are aggregating this data and not collecting it on an individual basis - there is unfortunately no way you can prove or disprove that claim, since the public has no visibility into what code is running on your servers.  So users have to trust you, where before they did not (which was a central point of the design of the app).

Users also have to trust that there are no malicious insiders working at CDS-SNC (which, hopefully, there are not), who could easily intercept and log this data.

This data is also sent over the public internet.  It's encrypted with TLS, but are you *sure* no one passively listening to the encrypted stream can figure out whether or not a user has been shown a "You have been exposed to covid" alert?  (Hint: you shouldn't be - check the message I just left with your security team.)  Users have to trust that you've designed a secure system, and unfortunately there is no such thing; you can only ever design a system so secure that someone as smart as you can't break into it.  But there's always someone out there smarter...

> The voluntary nature of the app means there is always a possibility that app users will delete the app.

If users have no opt-out, then what you're effectively saying to users is "You have to trust these things, and if you don't, we're going to take our ball and go home."  All of the above are risks, and by not providing an opt-out, you have effectively decided for your users that these risks are acceptable.  Some of your users may disagree with this position, and may have a different tolerance for risk than you do.  And this app is not a fun color matching app, it's an app that helps keep people safe.

Those users (which I count myself among) are now in the extremely uncomfortable position of "Do I leave this app installed, and take a risk with my privacy, or do I uninstall it, and forego the protection it offers me?"

>  ...understand if people are choosing to delete the app.

Monitoring this is monitoring the people who weigh that decision and come down on the side of "I guess I value my privacy more than my safety".  But why are you forcing people to make this decision at all?  Just add an opt-out.

**Reply by obrien-j** [reply]
> > ...to be able to gather aggregate numbers
> 
> The data that's actually sent from my phone is a notification that I was shown, for example, a "You have been exposed to covid" alert, and my IP address. With my IP address, CDS-SNC receives all the information they need to track whom has and has not been shows such a notification. Users only have your word that you are aggregating this data and not collecting it on an individual basis - there is unfortunately no way you can prove or disprove that claim, since the public has no visibility into what code is running on your servers. So users have to trust you, where before they did not (which was a central point of the design of the app).

Just to note, the code for the [covid-alert-server](https://github.com/cds-snc/covid-alert-server) (and shortly, metrics endpoint [lambda]) is open for review, while also noting that proving that the code running in production is also in github is somewhat of a non-trivial problem to prove. 

We've taken great lengths to make our work open, and to include the Office of the Privacy Commisioner, and HC's Privacy management division to address this 'you have to trust' point, but I do accept the premise. I would argue that this same fundamental problem existed previously, in that in order to upload 'Diagnosis keys', indicating a postive diagnosis, our systems also collected the IP address for 3 months, as mentioned in our Privacy policy, with specific restrictions on usage.

> 
> This data is also sent over the public internet. It's encrypted with TLS, but are you _sure_ no one passively listening to the encrypted stream can figure out whether or not a user has been shown a "You have been exposed to covid" alert? (Hint: you shouldn't be - check the message I just left with your security team.) Users have to trust that you've designed a secure system, and unfortunately there is no such thing; you can only ever design a system so secure that someone as smart as you can't break into it. But there's always someone out there smarter...

I lead security at CDS, and previously spent 12 years in Cyber Defence @ CSE. I am with you on the whole passive interception and person on the side inference attacks, it equally applies to the [existing upload for diagnosis keys](https://github.com/cds-snc/covid-alert-app/issues/986). This is an outstanding issue that we are aware of, but have not yet had the priority cycles to address. We have to weigh the resources we have with the goals of the service and make decisions accordingly.

> > The voluntary nature of the app means there is always a possibility that app users will delete the app.
> 
> If users have no opt-out, then what you're effectively saying to users is "You have to trust these things, and if you don't, we're going to take our ball and go home." All of the above are risks, and by not providing an opt-out, you have effectively decided for your users that these risks are acceptable. Some of your users may disagree with this position, and may have a different tolerance for risk than you do. And this app is not a fun color matching app, it's an app that helps keep people safe.
> 
> Those users (which I count myself among) are now in the extremely uncomfortable position of "Do I leave this app installed, and take a risk with my privacy, or do I uninstall it, and forego the protection it offers me?"

In it's current form, yes, that is effectively the choice you have, and one that I can not make on your behalf. I can only state that as designed, I do not believe the risk outweighs the benefit to Canadians, and have no issues installing this on my or my extended families devices.

> 
> > ...understand if people are choosing to delete the app.
> 
> Monitoring this is monitoring the people who weigh that decision and come down on the side of "I guess I value my privacy more than my safety". But why are you forcing people to make this decision at all? Just add an opt-out.

Thanks again for raising the issue, I do understand and respect your concerns. I've raised the issue with the broader team again, and will raise it with our partners at HC.

**Reply by jwalton** [reply]
> We've taken great lengths to make our work open...

Definitely, and it's very much appreciated, and I certainly don't mean to belittle that.  Please don't take anything I've said as a criticism of that.  I only meant to point out that no matter how open you make the code, there's no sensible way you can let the average Canadian poke around in your AWS account.

> In order to upload 'Diagnosis keys', indicating a postive diagnosis, our systems also collected the IP address for 3 months

The Diagnosis key is a little different though - if I receive a positive diagnosis, I can "opt out" by simply never entering my key into the app.  I have to interact with the app in order for this data to be collected, and it's clear that my one time key is going to be sent to a server along with my exposure codes, so it's clear there some data about me being sent somewhere.

Prior to the new metrics feature, if I was shown an exposure notification, though, only my device and I knew there was a possibility I'd been exposed - this information never left my device.  Now it does, and importantly it does so without any interaction from me.  The only way for a user to stop it being sent out is to uninstall the app and hope they do it fast enough that they do it before the metrics collection runs.

**Reply by obrien-j** [reply]
> > We've taken great lengths to make our work open...
> 
> Definitely, and it's very much appreciated, and I certainly don't mean to belittle that. Please don't take anything I've said as a criticism of that. I only meant to point out that no matter how open you make the code, there's no sensible way you can let the average Canadian poke around in your AWS account.

Not at all. This is all fair game imho.

> > In order to upload 'Diagnosis keys', indicating a postive diagnosis, our systems also collected the IP address for 3 months
> 
> The Diagnosis key is a little different though - if I receive a positive diagnosis, I can "opt out" by simply never entering my key into the app. I have to interact with the app in order for this data to be collected, and it's clear that my one time key is going to be sent to a server along with my exposure codes, so it's clear there some data about me being sent somewhere.
> 
> Prior to the new metrics feature, if I was shown an exposure notification, though, only my device and I knew there was a possibility I'd been exposed - this information never left my device. Now it does, and importantly it does so without any interaction from me. The only way for a user to stop it being sent out is to uninstall the app and hope they do it fast enough that they do it before the metrics collection runs.

Fair enough on the 'consent' piece.

**Comment by jwalton** [comment]
For users wishing to use this app and prevent metrics collection, metrics are sent to a server at the address `metrics.covid-notification.alpha.canada.ca`.  You can block traffic to this server via [pihole](https://pi-hole.net/) or any other DNS based blocking system.

This is a little tricky to do, since we're talking about a mobile device here, and if your device leaves your house and disconnects from your VPN, you may find yourself on your carrier network and the covid-alert app will have a chance to send metrics.  Also, note that a causal perusal of the source suggests that covid-alert will keep storing metrics until it has a chance to send them, so be aware that covid-alert may start slowly eating all the space on your device (although metrics are small, and hopefully covid will be over by the time it starts taking an appreciable amount of space).

**Reply by obrien-j** [reply]
I've flagged the potential for a storage space leak to the app team, who will likely put in a time based filter in the interim, while discussions are ongoing regarding opt-out status. 

Thanks. cc: @timarney
